plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/WebProjectUtil.java

```diff
@@ -70,4 +92,125 @@ public static IFile findInWebInf(IProject project, IPath filePath) {
     return file.exists() ? file : null;
   }
 
+  private static List<IFolder> findAllWebInfFolders(IContainer container) {
+    List<IFolder> webInfFolders = new ArrayList<>();
+    try {
+      for (IResource resource : container.members()) {
+        if (resource.exists() && resource.getType() == IResource.FOLDER) {
+          if ("WEB-INF".equals(resource.getName())) {
+            webInfFolders.add((IFolder) resource);
+          } else {
+            webInfFolders.addAll(findAllWebInfFolders((IFolder) resource));
+          }
+        }
+      }
+    } catch (CoreException ex) {}
+    return webInfFolders;
+  }
+
+  /**
+   * Returns the web content path set by the Dynamic Web Module facet. May return {@code null}.
+   *
+   * The path information resides in an XML file, which a project settings file:
+   * ".settings/org.eclipse.wst.common.component". This methods effectively returns the path
+   * tagged with {@code "defaultRootSource"} in the XML file.
+   */
+  private static IPath getWtpWebContentPathSetting(IProject project) {
+    IVirtualComponent component = ComponentCore.createComponent(project);
+    if (component != null && component.exists()) {
+      IVirtualFolder rootFolder = component.getRootFolder();
+      if (rootFolder instanceof ITaggedVirtualResource) {
+        ITaggedVirtualResource resource = (ITaggedVirtualResource) rootFolder;
+        IPath defaultRootSource = resource.getFirstTaggedResource(
+            WorkbenchComponent.DEFAULT_ROOT_SOURCE_TAG /* "defaultRootSource" String literal */);
+        return defaultRootSource;
+      }
+    }
+    return null;
+  }
+
+  public static boolean isDefaultWebContentFolderBogus(IProject project) {
```

Comment:
Singular (`...Folder`), I think. We want to detect if the Dynamic Web facet created a very specific directory `/WebContent`.